### PR TITLE
Finalise shareKeys

### PR DESCRIPTION
**Pull request by nf-s**
## Finalise `shareKeys`

Fixes #https://github.com/TerriaJS/terriajs/issues/4774

See https://github.com/TerriaJS/terriajs/blob/next/architecture/0005-root-group-ids-and-shareKeys.md for background

### V7 catalog support

Partially fixes https://github.com/TerriaJS/terriajs/issues/4897

This replaces https://github.com/TerriaJS/terrace/pull/158

Adds `v7initializationUrls` to terria config. It will convert catalogs to v8 and print messages to console.

For example:

```json
    {
      "v7initializationUrls": [
        "https://raw.githubusercontent.com/GeoscienceAustralia/dea-config/master/dev/terria/dea-maps.json",
        "https://maps.dea.ga.gov.au/init/terria-cube.json"
      ]
    },
```

### ShareKeys support for root group of Magda map config

To map `map-config-de-australia` to the root group (now `"/"`) you can add a `shareKey` like so:

```json
{
  "aspects": {
    "terria": {
      "id": "map-config-de-australia",
      "shareKeys": [
        "map-config-de-australia"
      ],
      "type": "group"
    },
    ...
  },
  "id": "map-config-de-australia",
  ...
}
```

### ShareKeys support for dynamic groups

See ADR for more info - https://github.com/TerriaJS/terriajs/blob/77a25b6967281e1f6189ccf54d77b2974730d4e1/architecture/0006-dynamic-groups-and-shareKeys.md

### Change how WMS group sets `id` for items

To mimic v7. See https://github.com/TerriaJS/terriajs/blob/77a25b6967281e1f6189ccf54d77b2974730d4e1/architecture/0006-dynamic-groups-and-shareKeys.md#wms-group

### Other small changes

* Added `InitDataPromise` to `InitSources`
* Add reverse `modelIdShareKeysMap` map - `model.id` -> `shareKeys`
* Reverted Legend use of `object` instead of `img` - sometimes it was showing html error responses
* Legend will now hide if an error is thrown

### Checklist

-   [x] Add test for magda map config shareKeys
-  [x] Add test for GroupMixin shareKeys
-  [x] Add test for v7 catalog
-   [x] I've updated CHANGES.md with what I changed.

### CODE REVIEW

File: lib/Models/InitSource.ts [hunk 1]
```diff
@@ -25,6 +27,15 @@ export function isInitData(initSource: InitSource): initSource is InitData {
...
 
+export function isInitDataPromise(
+  initSource: InitSource
+): initSource is InitDataPromise {
+  return (
+    initSource &&
+    Object.prototype.toString.call(initSource) === "[object Promise]"
+  );
+}
+
```
Comment (nf-s) [inline lib/Models/InitSource.ts:38]: Used for v7-v8 catalog converting

File: lib/Core/addedByUser.ts [hunk 2]
```diff
@@ -26,10 +28,9 @@ export default function addedByUser(
   return sourceReference.knownContainerUniqueIds.some(containerId => {
     return (
       containerId === USER_ADDED_CATEGORY_ID ||
-      addedByUser(
-        <BaseModel>catalogMember.terria.getModelById(BaseModel, containerId),
-        { depth: depth + 1 }
-      )
+      addedByUser(catalogMember.terria.getModelById(BaseModel, containerId), {
+        depth: depth + 1
+      })
```
Comment (nf-s) [inline lib/Core/addedByUser.ts:33]: Removed bad `<BaseModel>` typing here

File: lib/Models/Terria.ts [hunk 3]
```diff
@@ -1026,7 +1051,8 @@ export default class Terria {
       let existingReference = this.getModelById(MagdaReference, id);
       if (existingReference === undefined) {
         existingReference = new MagdaReference(id, this);
-        this.addModel(existingReference);
+        // Add model with terria aspects shareKeys
+        this.addModel(existingReference, aspects?.terria?.shareKeys);
```
Comment (nf-s) [inline lib/Models/Terria.ts:1020]: For magda root group sharekeys

File: lib/Models/Terria.ts [hunk 4]
```diff
@@ -1206,34 +1232,32 @@ function generateInitializationUrl(
...
-    return promise.then(jsonValue => {
-      if (isJsonObject(jsonValue)) {
-        return jsonValue;
-      }
-      return undefined;
-    });
+    if (jsonValue && isJsonObject(jsonValue)) {
+      return jsonValue;
+    }
+    return undefined;
```
Comment (nf-s) [inline lib/Models/Terria.ts]: Asyncified + added support for `InitDataPromise` - which is just a `Promise` which resolves to `InitData`

File: lib/Models/WebMapServiceCatalogGroup.ts [hunk 5]
```diff
@@ -288,7 +288,7 @@ class GetCapabilitiesStratum extends LoadableStratum(
     if (!isDefined(this.catalogGroup.uniqueId)) {
       return;
     }
-    return `${this.catalogGroup.uniqueId}/${layer.Name || layer.Title}`;
+    return `${this.catalogGroup.uniqueId}/${layer.Title}`;
```
Comment (nf-s) [inline lib/Models/WebMapServiceCatalogGroup.ts:291]: Revert WMS id to match v7

File: lib/ReactViews/Workbench/Controls/Legend.tsx [hunk 6]
```diff
@@ -49,17 +49,25 @@ export default class Legend extends React.Component<{
...
+    image.style.display = "initial";
+
+    // If legend need scaling, this is the only way to do it :/
+    // See https://stackoverflow.com/questions/7699621/display-image-at-50-of-its-native-size
+    // or https://stackoverflow.com/questions/35711807/display-high-dpi-image-at-50-scaling-using-just-css
+
+    image.style.width = `${(legend.imageScaling ?? 1) * image.offsetWidth}px`;
     // Must set maxWidth *after* setting width, as it may change offsetWidth
     image.style.maxWidth = "100%";
   }
```
Comment (nf-s) [inline lib/ReactViews/Workbench/Controls/Legend.tsx:73]: Refactored this to hide if an image request gives error response.

File: package.json [hunk 7]
```diff
@@ -58,7 +58,7 @@
     "@vx/tooltip": "^0.0.193-alpha.2",
     "babel-loader": "^8.0.5",
     "babel-plugin-jsx-control-statements": "^4.0.0",
-    "catalog-converter": "^0.0.2-alpha.3",
+    "catalog-converter": "^0.0.2-alpha.4",
```
Comment (nf-s) [inline package.json:61]: This depends on https://github.com/TerriaJS/catalog-converter/pull/32

**Comment by nf-s** [comment]
- [x] Before merging - merge https://github.com/TerriaJS/catalog-converter/pull/32 and publish new `catalog-converter` version

File: lib/ReactViews/Notification/shareConvertNotification.tsx [hunk 8]
```diff
@@ -11,11 +11,11 @@ const Spacing: any = require("../../Styled/Spacing").Spacing;
 const RawButton: any = require("../../Styled/Button").RawButton;
 const TextSpan: any = require("../../Styled/Text").TextSpan;
 
-export const shareConvertNotification = (shareProps: any) => (
+export const shareConvertNotification = (messages: any) => (
```
Comment (tephenavies) [inline lib/ReactViews/Notification/shareConvertNotification.tsx]: We have typings for this. Adding:
```suggestion
export const shareConvertNotification = (
  messages: NotUndefined<import("catalog-converter").ShareResult["messages"]>
) => (viewState: ViewState) => {
```
I realised that (according to typings at least) `messages` can be undefined. That should be handled in either Terria.ts or here.

Comment (tephenavies) [reply lib/ReactViews/Notification/shareConvertNotification.tsx]: Looks like it's always an Array in catalog-converter code. We should change the types in catalog-converter.

**Review by tephenavies — CHANGES_REQUESTED** [review]

File: lib/Models/Terria.ts (hunk 4 above)
Comment (tephenavies) [reply lib/Models/Terria.ts]: Still reviewing this

File: lib/Models/Terria.ts [hunk 9]
```diff
@@ -1206,34 +1197,32 @@ function generateInitializationUrl(
...
+          const json = await previousOptionPromise;
+          if (json === undefined) {
             return loadInitSource(option);
-          });
+          }
+          return json;
+        } catch (_) {
+          return loadInitSource(option);
+        }
       }, Promise.resolve<JsonObject | undefined>(undefined));
```
Comment (tephenavies) [inline lib/Models/Terria.ts]:
```suggestion
      let error: any;
      for (const option of initSource.options) {
        try {
          jsonValue = await loadInitSource(option);
          if (jsonValue !== undefined) break;
        } catch (err) {
          error = err;
        }
      }
      if (jsonValue === undefined && error !== undefined) throw error;
```

Comment (tephenavies) [reply lib/Models/Terria.ts]: I think that is more clear. It was a bit chaotic. The `error = err` but is still a bit weird. You could replace that with a true `TerriaError` with a title and message (i.e. just empty catch and then throw a new error at 1214 if `jsonValue === undefined`). We probably will at some point.

File: lib/Models/WebMapServiceCatalogItem.ts [hunk 10]
```diff
@@ -207,13 +207,16 @@ class GetCapabilitiesStratum extends LoadableStratum(
...
-              terriaTheme.textLight.split("#")[1]
-            }`;
+            let terriaTextLight = terriaTheme.textLight.split("#")[1];
+
+            // If terriaTextLight is a 3-character hex -> turn into 6
+            if (terriaTextLight.length === 3) {
+              terriaTextLight = `${terriaTextLight[0]}${terriaTextLight[0]}${terriaTextLight[1]}${terriaTextLight[1]}${terriaTextLight[2]}${terriaTextLight[2]}`;
+            }
+            legendOptions += `;fontColor:0x${terriaTextLight}`;
           }
```
Comment (nf-s) [inline lib/Models/WebMapServiceCatalogItem.ts]: If `textLight` is `#fff` - Geoserver gives us a blue legend ?!

Comment (nf-s) [reply lib/Models/WebMapServiceCatalogItem.ts]: @steve9164 give me a second, just need to fix this

File: lib/Models/WebMapServiceCatalogItem.ts [hunk 11]
```diff
@@ -207,13 +207,14 @@ class GetCapabilitiesStratum extends LoadableStratum(
...
+          let fontColor = terriaTheme.textLight.split("#")?.[1];
           if (isDefined(fontColor)) {
-            legendOptions += `;fontColor:0x${
-              terriaTheme.textLight.split("#")[1]
-            }`;
+            // If fontColor is a 3-character hex -> turn into 6
+            if (fontColor.length === 3) {
+              fontColor = `${fontColor[0]}${fontColor[0]}${fontColor[1]}${fontColor[1]}${fontColor[2]}${fontColor[2]}`;
+            }
+            legendOptions += `;fontColor:0x${fontColor}`;
```
Comment (nf-s) [inline lib/Models/WebMapServiceCatalogItem.ts:217]: If `textLight` is `#fff` - Geoserver gives us a blue legend ?!

**Review by tephenavies — APPROVED** [review]
Tests pass locally

**Pull request merged**